caas/kubernetes/provider/upgrade.go

```diff
@@ -17,19 +18,118 @@ import (
 	k8slabels "k8s.io/apimachinery/pkg/labels"
 	k8stypes "k8s.io/apimachinery/pkg/types"
 
+	"github.com/juju/juju/caas"
 	"github.com/juju/juju/cloudconfig/podcfg"
 	k8sannotations "github.com/juju/juju/core/annotations"
 	"github.com/juju/juju/core/status"
 )
 
 const applicationUpgradeTimeoutSeconds = 30
 
+type UpgradeBroker interface {
+	// Operator returns an Operator with current status and life details.
+	Operator(string) (*caas.Operator, error)
+
+	// Namespace returns the current default namespace targeted by this broker.
+	Namespace() string
+
+	StatefulSetForName(name) (*apps.StatefulSet, error)
+}
+
+type UpgradeCAASOperatorBroker interface {
+	// Returns the deployment name use for the given application name, supports
+	// finding legacy deployment names if set to True.
+	DeploymentName(string, bool) string
+
+	PatchStatefulSet(string, k8stypes.PatchType, []byte, ...string) (*apps.StatefulSet, error)
+
+	StatefulSetForName(name) (*apps.StatefulSet, error)
+}
+
+func upgradeAgent(agentTag string, vers version.Number) error {
+	tag, err := names.ParseTag(agentTag)
+	if err != nil {
+		return errors.Annotate(err, "parsing agent tag to upgrade")
+	}
+
+	name := tag.Id()
+	switch tag.Kind() {
+	case names.MachineTagKind:
```

Comment:
there should never be a machine tag for k8s

Reply:
This is just keeping the logic that was already there.